ts/util/reactions.ts

```diff
@@ -0,0 +1,156 @@
+    if (timestamp - firstTimestamp < rateTimeLimit) {
+      latestReactionTimestamps.pop();
+      return;
+    } else {
+      latestReactionTimestamps.shift();
+    }
+  }
+
+  const found = await getMessageById(messageId);
+  if (found && found.get('sent_at')) {
+    const conversationModel = found?.getConversation();
+    if (!conversationModel) {
+      window.log.warn(`Conversation for ${messageId} not found in db`);
+      return;
+    }
+
+    const me = UserUtils.getOurPubKeyStrFromCache();
+    const author = found.get('source');
+    let action = 0;
+
+    const reacts = found.get('reacts');
+    if (reacts && Object.keys(reacts).includes(emoji) && Object.keys(reacts[emoji]).includes(me)) {
+      window.log.info('found matching reaction removing it');
+      action = 1;
+    } else {
+      const reactions = await getRecentReactions();
+      if (reactions) {
+        await updateRecentReactions(reactions, emoji);
+      }
+    }
+
+    await conversationModel.sendReaction(messageId, {
+      id: Number(found.get('sent_at')),
+      author,
+      emoji,
+      action,
+    });
+
+    window.log.info(
+      me,
+      `${action === 0 ? 'added' : 'removed'} a`,
+      emoji,
+      'reaction at',
+      found.get('sent_at')
+    );
+  } else {
+    window.log.warn(`Message ${messageId} not found in db`);
+  }
+};
+
+/**
+ * Handle reactions on the client by updating the state of the source message
+ */
+export const handleMessageReaction = async (
+  reaction: SignalService.DataMessage.IReaction,
+  sender: string,
+  messageId?: string
+) => {
+  const originalMessageTimestamp = Number(reaction.id);
+  const originalMessageAuthor = reaction.author;
+
+  if (!reaction.emoji) {
+    window?.log?.warn(`There is no emoji for the reaction ${messageId}.`);
+    return;
+  }
+
+  const collection = await getMessagesBySentAt(originalMessageTimestamp);
+  const originalMessage = collection.find((item: MessageModel) => {
+    const messageTimestamp = item.get('sent_at');
+    const author = item.get('source');
+    return Boolean(
+      messageTimestamp &&
+        messageTimestamp === originalMessageTimestamp &&
+        author &&
+        author === originalMessageAuthor
+    );
+  });
+
+  if (!originalMessage) {
+    window?.log?.warn(`We did not find the original reacted message ${originalMessageTimestamp}.`);
+    return;
+  }
+
+  const reacts: ReactionList = originalMessage.get('reacts') ?? {};
+  reacts[reaction.emoji] = reacts[reaction.emoji] || {};
+  const details = reacts[reaction.emoji] ?? {};
+  const senders = Object.keys(details);
+
+  switch (reaction.action) {
+    // Add reaction
+    case 0:
+      if (senders.includes(sender) && details[sender] !== '') {
+        window?.log?.info('Received duplicate message reaction. Dropping it. id:', details[sender]);
+        return;
+      }
+      details[sender] = messageId ?? '';
+      break;
+    // Remove reaction
+    case 1:
```

Comment:
you can use the enum type raher than their int value
something like
`SignalService.DataMessage.Reaction.Action.REACT`

Reply:
nice idea!